tools/biosplitmerge.c

```diff
@@ -0,0 +1,192 @@
+/* 
+ * biosplitmerge.c      The eBPF C source file for split and merge tracer. 
+ *                      Functions in this file are imported and used by 
+ *                      biosplitmerge.py. 
+ *
+ *
+ * Copyright (c) Google LLC
+ * Licensed under the Apache License, Version 2.0 (the "License")
+ *
+ * 09-Aug-2020   Haoning Chen   Created this.
+ */
+
+#include <uapi/linux/ptrace.h>
+#include <linux/blkdev.h>
+#include <linux/version.h>
+
+/* Split/Merge event type. */
+enum type { Split = 0, Fmerge, Bmerge, Dmerge };
+
+
+/* Value of the bpf map */
+struct val_t {
+	u64 ts; 
+	u64 sector;
+	u64 len;
+	u64 type;
+};
+
+
+/* Output data type sent from kernel to user space */
+struct data_t {
+	u64 ts; 
+	u64 in_sector;
+	u64 in_len;
+	u64 out_sector; 
+	u64 out_len; 
+	u64 rwflag; 
+	u64 type;
+	u64 pid;
+	char disk_name[DISK_NAME_LEN];
+	char cmd_name[TASK_COMM_LEN];
+};
+
+
+BPF_HASH(input, u64, struct val_t);    // use pid as hash key
+BPF_PERF_OUTPUT(events); 
+
+
+/* Common actions performed when probe function is entered. */
+static inline void do_entry(enum type event_type, struct bio *bio) {
+	u64 pid = bpf_get_current_pid_tgid();
+	// put the input bio's data to map 
+	struct bvec_iter bi_iter;
+	bpf_probe_read_kernel(&bi_iter, sizeof(bi_iter), &bio->bi_iter); 
+
+	struct val_t val = { 
+		.ts     = bpf_ktime_get_ns(), 
+		.sector = bi_iter.bi_sector, 
+		.len    = bi_iter.bi_size,
+		.type   = (u64) event_type
+	}; 
+	input.update(&pid, &val);
+}
+
+
+/* Function probed to entry of bio_split(). */
+int split_entry(struct pt_regs *ctx, struct bio *bio) {
+	do_entry(Split, bio);
+	return 0;
+}
+
+
+/* Function probed to entry of bio_attempt_front_merge(). */
+int front_merge_entry(struct pt_regs *ctx) {
+	// obtain the second param of the function
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(5, 3, 0)
+	struct bio *bio = (struct bio *)PT_REGS_PARM2(ctx);
+#else 
+	struct bio *bio = (struct bio *)PT_REGS_PARM3(ctx);
+#endif
+	do_entry(Fmerge, bio);
+	return 0;
+}
+
+/* Function probed to entry of bio_attempt_back_merge(). */
+int back_merge_entry(struct pt_regs *ctx) {
+	// obtain the bio argument of the function
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(5, 3, 0)
+	struct bio *bio = (struct bio *)PT_REGS_PARM2(ctx);
+#else 
+	struct bio *bio = (struct bio *)PT_REGS_PARM3(ctx);
+#endif
+	do_entry(Bmerge, bio);
+	return 0;
+}
+
+
+/* Function probed to entry of bio_attempt_discard_merge(). */
+int discard_merge_entry(struct pt_regs *ctx) {
+	struct bio *bio = (struct bio *)PT_REGS_PARM3(ctx);
+	do_entry(Dmerge, bio);
+	return 0;
+}
+
+
+/* Function probed to return of bio_split(). */
+int split_return(struct pt_regs *ctx) {
+	u64 pid = bpf_get_current_pid_tgid();
+	struct bio *split = (struct bio *)PT_REGS_RC(ctx); 
+	
+	struct val_t *valp = input.lookup(&pid); 
+	if (valp == NULL) 
+		return 0; 
+
+	// verify the split event and construct output data struct
+	if (split != NULL) {
```

Comment:
I'm just not used to using "goto", since at school it is strictly banned using "goto" statement lol. I'll delete the map entry right after the look up and return.